src/Tasks/Microsoft.NET.Build.Tasks/ResolveAppHosts.cs

```diff
@@ -0,0 +1,166 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text;
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+using NuGet.Frameworks;
+
+namespace Microsoft.NET.Build.Tasks
+{
+    public class ResolveAppHosts : TaskBase
+    {
+        public string TargetFrameworkIdentifier { get; set; }
+
+        public string TargetFrameworkVersion { get; set; }
+
+        public string TargetingPackRoot { get; set; }
+
+        public string AppHostRuntimeIdentifier { get; set; }
+
+        public ITaskItem[] PackAsToolShimRuntimeIdentifiers { get; set; } = Array.Empty<ITaskItem>();
+
+        /// <summary>
+        /// The file name of Apphost asset.
+        /// </summary>
+        [Required]
+        public string DotNetAppHostExecutableNameWithoutExtension { get; set; }
+
+        [Required]
+        public string RuntimeGraphPath { get; set; }
+
+        public ITaskItem[] KnownAppHostPacks { get; set; }
+
+        [Output]
+        public ITaskItem[] PackagesToDownload { get; set; }
+
+        //  There should only be one AppHost item, but we use an item here so we can attach metadata to it
+        //  (ie the apphost pack name and version, and the relative path to the apphost inside of it so
+        //  we can resolve the full path later)
+        [Output]
+        public ITaskItem[] AppHost { get; set; }
+
+        [Output]
+        public ITaskItem[] PackAsToolShimAppHostPacks { get; set; }
+
+        protected override void ExecuteCore()
+        {
+            var knownAppHostPacksForTargetFramework = KnownAppHostPacks
+                .Where(appHostPack =>
+                {
+                    var packTargetFramework = NuGetFramework.Parse(appHostPack.GetMetadata("TargetFramework"));
+                    return packTargetFramework.Framework.Equals(TargetFrameworkIdentifier, StringComparison.OrdinalIgnoreCase) &&
+                        ResolveFrameworkReferences.NormalizeVersion(packTargetFramework.Version) ==
+                            ResolveFrameworkReferences.NormalizeVersion(new Version(TargetFrameworkVersion));
+                })
+                .ToList();
+
+            if (knownAppHostPacksForTargetFramework.Count > 1)
+            {
+                throw new InvalidOperationException("Multiple KnownAppHostPack items applied to the specified target framework, which is not supposed to happen");
```

Comment:
I'm torn on this being a non-BuildErrorException because KnownAppHostPack could be modified by anyone. Unlikely though, and I guess this is just moved from elsewhere.

Reply:
I feel like it's OK to throw these types of errors when the invalid input is something that wouldn't be public if there was a concept of non-public items in MSBuild.

Reply:
Should we put "_" in front of it if it wouldn't be public?